File: heap/0659_split_array_into_consecutive_subsequences/0659_split_array_into_consecutive_subsequences.go
package main 

import (
	"fmt"
)

func isPossible(nums []int) bool {
	counter := make(map[int]int)
	tail := make(map[int]int)
	for i := 0; i< len(nums); i++ {
		counter[nums[i]]++
	}

	for _, v := range nums {
		if counter[v] == 0{
			continue
		}

		if n := tail[v-1]; n> 0 {
			tail[v-1]--
			tail[v]++
			counter[v]--
			continue
		}
		if counter[v+1] > 0 && counter[v+2] > 0 {
			tail[v+2]++
			counter[v]--
			counter[v+1]--
			counter[v+2]--
		} else {
			return false
		}
	}
	return true
}

func main() {
	nums := []int{1,2,3,3,4,5}
	res := isPossible(nums)
	fmt.Println(res)
}

File: heap/0659_split_array_into_consecutive_subsequences/0659_split_array_into_consecutive_subsequences.py

import collections
import heapq


class Solution(object):
    def isPossible(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        saved = collections.defaultdict(list)
        for num in nums:
            last = saved[num - 1]
            _len = 0 if (not last) else heapq.heappop(last)
            current = saved[num]
            heapq.heappush(current, _len + 1)
        for values in saved.values():
            for v in values:
                if v < 3:
                    return False
        return True


nums = [1,2,3,3,4,5]
res = Solution().isPossible(nums)
print(res)